### DISCUSSION TITLE
Code coverage for Haraka plugin tests?

### DISCUSSION THREAD

**Discussion by tabascoterrier** [haraka/Haraka#3218] [Q&A] [answered]
I'm trying to do better at testing my plugins. Is there a way of getting code coverage reporting working with Haraka plugin tests?

Using the plugin template as an example:

- git clone https://github.com/haraka/haraka-plugin-template.git
- npm install nyc --save-dev
- Update the `package.json` test command to be `"test": "nyc --all npx mocha"`
 
If I run the tests they pass fine, but no coverage is calculated. I'm quite a nodejs noob, but I'm guessing this is due to the plugin code being loaded via `haraka-test-fixtures` and not directly, and nyc isn't picking up on that?  If so, is there an easy way around this?

> By default, nyc only collects coverage for source files that are visited during a test. It does this by watching for files that are require()'d during the test. When a file is require()'d, nyc creates and returns an instrumented version of the source, rather than the original. 

```
❯ npm test

> haraka-plugin-template@1.0.3 test
> nyc --all npx mocha
  template
    ✔ loads

  load_template_ini
    ✔ loads template.ini from config/template.ini
    ✔ initializes enabled boolean

  uses text fixtures
    ✔ sets up a connection
    ✔ sets up a transaction


  5 passing (7ms)

----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
----------|---------|----------|---------|---------|-------------------
All files |       0 |      100 |       0 |       0 |
 index.js |       0 |      100 |       0 |       0 | 3-18
----------|---------|----------|---------|---------|-------------------
```

**Comment by msimerson** [comment, marked as answer]
> I'm trying to do better at testing my plugins.

🎉❤️

>  Is there a way of getting code coverage reporting working with Haraka plugin tests?

Unfortunately, not without ripping all the vm.runInContext stuff out of Haraka. That's what is interfering with the testing coverage. PRs are welcome.

**Reply by tabascoterrier** [reply]
Ah, fair enough. That kind of thing is beyond my capabilities at the moment unfortunately, but my node-fu is slowly improving so maybe I'll have a bash at that one day. Thanks for the speedy response!
